### DISCUSSION TITLE
Behavioral Modelling of Phase-Locked Loop using Deep Learning Techniques (Project 202)

### DISCUSSION THREAD

**Discussion by robertogl** [mathworks/MATLAB-Simulink-Challenge-Project-Hub#32] [Collaborate]
Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project #202](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Behavioral%20Modelling%20of%20Phase-Locked%20Loop%20using%20Deep%20Learning%20Techniques)

---
*:robot: An AI bot may reply to your comments with suggestions and guidance. It will not solve the project for you. Your comments may be processed by a third-party LLM. To direct a question to a specific person, tag them with @username.*

**Comment by lulf0020** [comment]
Hello, we are a team of undergraduates focusing on project #202 these days.  We now want to use the Mixed-Signal-Blockset, but we don't really know how to setup a Simulation-In-Loop to gather datas. Where can we find some help documents or references?  Thanks.

**Reply by pragatikt** [reply]
Some background on usage of PLL test-bench in our case here: You create a test model where PLL block is connected to a PLL testbench from Mixed-Signal Blockset. After the end of simulation, the results are saved in the model workspace . For accessing the document on how to access data from Model workspace programmatically you should type "doc modelworkspace" on MATLAB prompt. 
So, Simulation in  a loop in this case refers to a script where you can programmatically configure PLL (using for loop and 'set_param' command) and use programmatic way of accessing model workspace for getting the results (after getting information from 'doc modelworkspace'). 
Try this approach and hopefully you should be able to get the test data more efficiently for deep learning model.
Do, let me know if you have any other query.

-Pragati

**Comment by lulf0020** [comment]
Thanks for your help in "Simulation In Loop". Now when changing parameters, we find some problems. 
First, we don't know what effect param "Current imbalance " in the BLOCK "Integer N PLL with Dual Modulus Prescaler" will have on the system. Should we change "Current imbalance " as we are changing "leakage current" ?
Second, we don't find the parameter corresponding to "Input current" mentioned in the project. Is it referred to the param "Output current" in the BLOCK "Integer N PLL with Dual Modulus Prescaler/charge pump" ? 
I'd appreciate it if you could offer some help.

**Reply by pragatikt** [reply]
For a robust deep learning model you should have PLL configuration with varying leakage current only, varying charge imbalance only, and for some cases varying them both at the same time. 
Also, you are right about output current parameter in charge pump. 
HTH
Pragati

**Comment by lulf0020** [comment]
Excuse me.
we now have some problems, and may need some help.

1. About PLL dataset. Although we are collecting dataset from the Simulink, it's a bit slow. We wonder if there is pre-collected dataset which we can download.

2.About Deep Network Designer. When we want to import datastore to train the model, it show out "There are no data storage variables in the workspace". We don't know if it has data format requirements, and the help document online is not detailed enough.

3.About PLL model Simulation.When we try to change the parameter, it sometimes show warnings and errors as below:

- 1)"Input buffer overflow in Integer N PLL with Single Modulus Prescaler1/Loop Filter/Convert Sample Time due to closely spaced samples. Increase buffer size, possibly in a subsystem at a higher level."

We try to increase the buffer, but it doesn’t seem to solve the problem fundamentally. 

- 2)"Warning: The model was run for less than the recommended simulation stop time. Results reported might be inaccurate. For correct results run the simulation with a stop time >= minimum recommended simulation stop time of 0.00018 seconds. 
Warning: The simulation did not run long enough to generate necessary data for Phase noise measurement. Hence, some result entries might be 'NaNs'.
Run the simulation for the recommended simulation stop time provided in the testbench mask. 
(ERROR)
Propagation delay in Integer N PLL with Single Modulus Prescaler1/VCO/Variable Pulse Delay decreased more than variable step discrete input sample time increased, causing the output sample times to occur out of sequence."

For the Warnings above, we did change the simulation stop time to 0.0002s, but the warning still remain.
For the ERROR,  actually, we don't know how it occur. The only thing we know about it is that it occurs after the warnings above. 



I'd appreciate it if you could offer some help.

**Reply by pragatikt** [reply]
Q1: About PLL dataset. Although we are collecting dataset from the Simulink, it's a bit slow. We wonder if there is pre-collected dataset which we can download.

Answer: Unfortunately, we don't have any pre-collected dataset to share. So, you will have to generate the training data by running multiple simulations in a loop. I agree it might be time consuming.

With regards to warning and error messages, it seems that you are setting up a PLL with a configuration which is not feasible in real world. So, I might suggest to discard those cases when running simulations in the loop (instead of figuring out a way to correct them).
Hope that's helpful. 

-Pragati

**Comment by lulf0020** [comment]
I am sorry to bother you. Now we are dealing with the model and find a few questions/problems.

What bother us most is the PHASE Noise, an example is like the picture below. The wave is jagged. And at the beginning, we decide to use some sampling point in the frequency curve to represent the PHASE Noise characteristics, but now we find that because of the jagged curve, the tendency of the sampling point(we now choose 4 point) may be hard to decribe with Nerual Network(Or maybe our neural network needs further optimization?)

-As mentioned above, we want to ask that HOW to understand the target metrics "PHASE Noise"?  Should we use several sampling points to represent this metric or should we learn its overall curve and trend?(Maybe an overall curve is hard to describe in the model.)
Hope you could give us some suggestion!

![D2)1N2MXH 9MV2316`_1M(W](https://user-images.githubusercontent.com/88016174/141745478-f46faab2-1f92-4d31-a947-47cdc9451387.png)
r

Thanks!

**Reply by pragatikt** [reply]
You should not be worrying about figuring out target metric numbers. Avoid them. The important thing for your model should be the measured phase noise that the deep learning model should try to capture . Hope this is helpful.

**Comment by SiriRanganath** [comment]
what deep learning blockset to use? can anybody please help.

**Reply by pragatikt** [reply]
We encourage and recommend students to make use of deep learning toolbox by MathWorks. Following link will help you get more information: https://www.mathworks.com/help/deeplearning/

**Comment by dnoiro** [comment]
Hi! Are we meant to design the PLL simulation in loop from figure 2 or make our own?

**Reply by pragatikt** [reply]
It's recommended to use the testbench setup as illustrated in figure 2 to get the required metrics.

**Reply by dnoiro** [reply]
ok thx i got my training data. again with the figures, figure 3 shows the network only have 3 inputs, should I only use 3 inputs or all 6?

**Reply by pragatikt** [reply]
The image in figure 3 is just a representation of the kind of model one should strive to create using deep learning. The exact size shouldn't be replicated. The output of a PLL also depends on the loop filter configuration, so apart from Icp, Kvco and N, one should also consider putting in loop filter components as inputs. For proof of concept, you can start with 3 inputs as shown in figure and then grow that model once you have gain confidence. My 2 cents.

**Comment by whoisdinanath** [comment]
We are planning to work on this project and need some help.

**Reply by pragatikt** [reply]
Great to know you are interested in this project. Please drop your question in this forum and we will be happy to respond.

**Comment by Shika-Somireddy** [comment]
Hi, my name is Rishika and I am planning on doing this project for the 2025 mathworks AI challenge. Right now, I am reading some of the literature surrounding PLLs!

**Reply by pragatikt** [reply]
Welcome Rishika, and good luck !
